compiler/src/dotty/tools/dotc/transform/sjs/ExplicitJSClasses.scala

```diff
@@ -12,54 +17,728 @@ import core.Types._
+        if (ref1.is(NoInits)) {
+          // If one of the decls is a JS class, there is now some initialization code to create the JS class
+          if (!isJSType && ref1.info.decls.exists(isJSClass))
+            ref1.copySymDenotation(initFlags = ref1.flags &~ NoInits)
+          else
+            ref1
+        } else {
+          // JS traits never have an initializer, no matter what dotc thinks
+          if (isJSType)
+            ref1.copySymDenotation(initFlags = ref1.flags | NoInits)
+          else
+            ref1
+        }
+      case ref1 =>
+        ref1
+    }
   }
 
   override def infoMayChange(sym: Symbol)(using Context): Boolean =
     sym.isClass && !sym.is(JavaDefined)
 
-  override def transformTypeApply(tree: TypeApply)(using Context): tpd.Tree = {
-    tree match {
-      case TypeApply(fun, tpt :: Nil) if fun.symbol == jsdefn.JSPackage_constructorOf =>
-        ref(jsdefn.Runtime_constructorOf).appliedTo(clsOf(tpt.tpe))
+  override def prepareForUnit(tree: Tree)(using Context): Context =
+    ctx.fresh.updateStore(MyState, new MyState())
+
+  /** Populate `nestedObject2superClassTpe` for inner objects at the start of
+   *  a `Block` or `Template`, so that they are visible even before their
+   *  definition (in their enclosing scope).
+   */
+  private def populateNestedObject2superClassTpe(stats: List[Tree])(using Context): Unit = {
+    for (stat <- stats) {
+      stat match {
+        case cd @ TypeDef(_, rhs) if cd.isClassDef && cd.symbol.is(ModuleClass) && isInnerOrLocalJSClassOrObject(cd.symbol) =>
+          myState.nestedObject2superClassTpe(cd.symbol) = extractSuperTpeFromImpl(rhs.asInstanceOf[Template])
+        case _ =>
+      }
+    }
+  }
+
+  override def prepareForBlock(tree: Block)(using Context): Context = {
+    populateNestedObject2superClassTpe(tree.stats)
+    ctx
+  }
+
+  override def prepareForTemplate(tree: Template)(using Context): Context = {
+    populateNestedObject2superClassTpe(tree.body)
+    ctx
+  }
+
+  // This method implements steps (A) and (B)
+  override def transformTemplate(tree: Template)(using Context): Tree = {
+    val cls = ctx.owner.asClass
+
+    /* The `parents` of a Template have the same trees as `new` invocations
+     * of the parent classes and traits. That means that `transformApply` may
+     * have wrapped them in a `withContextualJSClassValue`, not knowing where
+     * they belong in the larger tree.
+     * We now unwrap those, canceling out that effect.
+     * TODO Is there a better way to do this?
+     */
+    val fixedParents =
+      if (!cls.isJSType) tree.parents // fast path
+      else tree.parents.mapConserve(unwrapWithContextualJSClassValue(_))
+
+    if (!isApplicableOwner(cls)) {
+      if (fixedParents eq tree.parents) tree
+      else cpy.Template(tree)(parents = fixedParents)
+    } else {
+      val newDecls = List.newBuilder[Tree]
+      for (decl <- tree.body) {
+        val declSym = decl.symbol
+        if (declSym eq null) {
```

Comment:
@smarter / @nicolasstucki I blindly copied this from scalac. Is there a better way to rule out non-member-defs? Maybe I should first pattern-match on `TypeDef`s, actually?

Reply:
All declarations have symbols. But the `body` of a template can also contain statements ;)

Anyway, I'll try the pattern match. It should be clearer.